packages/grafana-ui/src/types/panelOptions.tsx

```diff
@@ -0,0 +1,181 @@
+import { SelectOptionItem } from '../components/Select/Select';
+
+/**
+ * Option editors, i.e. BooleanOption have to comply to OptionInputAPI.
+ * OptionInputAPI represents UI elements that allow value modification, name it string, boolean, complex object, whatever
+ */
+
+export interface OptionInputAPI<T> {
+  value: T;
+  onChange: (value: T, event?: React.SyntheticEvent<HTMLElement>) => void;
+}
+
+export interface OptionsRowModel<TOptions> {
+  columns: number;
+  content: Array<OptionUIModel<TOptions, any> | GroupLayoutUIModel<any>>;
+}
+export interface OptionsUIModel<TOptions> {
+  rows: Array<OptionsRowModel<TOptions>>;
+}
+
+export enum OptionType {
+  Text = 'text',
+  Float = 'float',
+  Integer = 'integer',
+  Boolean = 'boolean',
+  Object = 'object',
+  Array = 'array',
+}
```

Comment:
I don't fully understand how this will work, but Array strikes me as a differnt camp.  I would expect somethign like:
DataType
* text
* float
* integer
* boolean
* custom (=> object)

Properties:
* multi-valued (stored as array)
* min/max/step (numbers)
* MaxLength (strings)
etc

Reply:
@ryantxu it is not used yet. Array/Object will not be there for sure. Thing is I wanted to use it for primitive option types to automatically select UI component based on the option type, if no component was specified. That's why `component` is an optional prop in `OptionUIModel` interface